### PR TITLE
Bug fix: Zero-length partitions result in NaN for overall mean & variance

### PR DESCRIPTION
**Pull request by ctn**
In the current code, when two merging partitions happen to have zero-length, the return mean will be NaN. Consequently, the result of mean or variance after reducing over all partitions will also be NaN, which is not correct if there are other partitions with non-zero length. This patch fixes this issue.

### CODE REVIEW

**Comment by mateiz** [comment]
Jenkins: this is ok to test

**Comment by rxin** [comment]
Christopher, can you add a test case for this?

**Comment by ctn** [comment]
Reynold, will do.

## 

Christopher T. Nguyen
L: http://linkedin.com/in/ctnguyen
C: http://doodle.com/cctn

**Comment by mateiz** [comment]
Any update on this -- do you guys have time to add a test?

**Comment by ctn** [comment]
ETA eow, Matei.

## 

Christopher T. Nguyen
L: http://linkedin.com/in/ctnguyen
C: http://doodle.com/cctn

**Comment by ctn** [comment]
Verified that running this test on old StatCounter gives:

```
[error] Test spark.JavaAPISuite.zeroLengthPartitions failed: expected:<6.0> but was:<NaN>
...
[error] Failed: : Total 35, Failed 1, Errors 0, Passed 34, Skipped 0
[error] Failed tests:
[error]     spark.JavaAPISuite
```

while running it on new StatCounter gives:

```
[info] Passed: : Total 35, Failed 0, Errors 0, Passed 35, Skipped 0
[success] Total time: 21 s, completed Jun 15, 2013 5:28:36 PM
```

**Comment by rxin** [comment]
Thanks for adding the test. Is there any reason you are adding this as a Java API Suite test? The Java API Suite is for testing the Java API only. It would be better to put this in a scala suite.

**Comment by ctn** [comment]
Good question, and no good reason other than applying the Levenshtein distance to StatCounter and finding JavaAPISuite.java. 

I'll move it to PartitioningSuite.scala.

**Comment by ctn** [comment]
Verified

```
> test-only spark.PartitioningSuite
...
[info] - Zero-length partitions should be correctly handled
...
[info] Passed: : Total 6, Failed 0, Errors 0, Passed 6, Skipped 0
[success] Total time: 10 s, completed Jun 16, 2013 1:29:23 AM
```

and

```
> test-only spark.JavaAPISuite
...
[info] Passed: : Total 34, Failed 0, Errors 0, Passed 34, Skipped 0
[success] Total time: 17 s, completed Jun 16, 2013 1:21:59 AM
```

File: project/plugins.sbt [hunk 1]
```diff
@@ -16,3 +16,5 @@ addSbtPlugin("io.spray" %% "sbt-twirl" % "0.6.1")
 //resolvers += Resolver.url("sbt-plugin-releases", new URL("http://scalasbt.artifactoryonline.com/scalasbt/sbt-plugin-releases/"))(Resolver.ivyStylePatterns)
 
 //addSbtPlugin("com.jsuereth" % "xsbt-gpg-plugin" % "0.6")
+
+libraryDependencies += "com.novocode" % "junit-interface" % "0.10-M4" % "test"
```
Comment (rxin) [inline project/plugins.sbt:20]: What is the use of this plugin?

Comment (ctn) [inline project/plugins.sbt:20]: I did a "git commit -a" and unintentionally included this, but you may wish
to leave that in. It makes it possible for "sbt/sbt test" to pick up Java
unit tests, e.g., JavaAPISuite.

## 

Christopher T. Nguyen
L: http://linkedin.com/in/ctnguyen
C: http://doodle.com/cctn

Comment (mateiz) [inline project/plugins.sbt:20]: We already have this plugin in SparkBuild.scala actually. Are you sure it's necessary?

Comment (mateiz) [inline project/plugins.sbt:20]: Actually, I'll merge this and remove it myself, seems simpler.

**Pull request merged**